File: Leetcode/Sliding Window Median/sliding window median.cpp
class Solution {
public:
    vector<double> medianSlidingWindow(vector<int>& nums, int k) {
        int a=0;
        int j=0;
        vector <double> res; // Declare result vector
        multiset <long long int, greater <int>> temp; //Multisets are a type of associative containers similar to set, with an exception that multiple elements can have same values.
        multiset <long long int, greater <int> > :: iterator itr; // declare iterator for the declared multiset to iterate through the data structure.
        for (int i=0;i<nums.size();i++){//iterate through provided vector.
            temp.insert(nums[i]);
            j++;
            if (j==k){ // when j reaches the window size.
                itr=temp.begin();
                advance(itr, temp.size() / 2);

                if(k%2 == 0) {
                    res.push_back((*itr + *(--itr))/2.0);  //even number: hence take average of both numbers/values lying in the middle.
                }else {
                    res.push_back(*itr); //for odd number: directly push the middle value(k/2)
                }
                temp.erase(temp.find(nums[a])); //erase the first number of the window.
                a++;
                j=k-1; // reduce the size of window by 1
            }
        }
        return res;
    }
};

File: Leetcode/Sliding Window Median/sliding window median.py
#480. Sliding Window Median
#https://leetcode.com/problems/sliding-window-median/
class Solution:
    def medianSlidingWindow(self, nums: List[int], k: int) -> List[float]:
        even = (k % 2 == 0)

        def median(even, nums, r, l):
            #Sort the array for selected window elements from left to right +1
            n = sorted(nums[l:r + 1])
            if even:
                #Pick average of 2 middle values
                med = (n[(k // 2) - 1] + n[k // 2]) / 2
            else:
                #When array odd pick middle value 
                med = n[k // 2]
            return med

        #Set position for left and right as zero for k elements of sliding window 
        r = 0
        l = 0
        ans = []
        #Iterate till we reach end of list
        while r < len(nums):
            #select k elements till it reaches k
            if r - l + 1 == k:
                 #Add median value for every slicing window till it reaches k slicing windows 
                ans.append(median(even, nums, r, l))
                l += 1
            r += 1
        return ans
